### PR TITLE
#8157: fix insert at cursor for draftjs editors

### LINKED ISSUE #8157: Insert at Cursor at end of text breaks DraftJS editors (editor disappears)
## Describe the bug

- On our sandbox, the editor disappears
- On client environment ([Kustomer](https://www.kustomer.com/) email reply editor on MV2), we're seeing a couple of issues on 1.8.11: 
  - The end of the text gets inserted 2x. The beginning of the text is missing
  - The field sometimes shows an error (it gets into an invalid state)

## To Reproduce

1. Visit https://pbx.vercel.app/advanced-fields/
2. Add some text to Draft JS editor and start a new line
3. Use Insert at Cursor from the sidebar

## Actual behavior

Editor dissapears on our playground

image

On client environment, when inserting at end of the line it's also inserting 2x

image

## Expected behavior

Text inserts at the end of the editor

## Additional Information
image

- See: `@pixies/template/snippet-manager-db`

## Workaround

- For the customer, inserting in the middle of existing text works OK. So it appears to be a problem with text boundaries/new elements in DraftJS

## Discussion

- We have custom logic for DraftJS in our Set Field Value brick: https://github.com/pixiebrix/pixiebrix-extension/blob/c68d199f9acaaceefcf03bc317f1e44f9e861044/src/utils/domFieldUtils.ts#L48-L48. 
- The check might not be correct: But it might not be getting toggled properly?: https://github.com/pixiebrix/pixiebrix-extension/blob/c68d199f9acaaceefcf03bc317f1e44f9e861044/src/contentScript/textEditorDom.ts#L82-L82. It may need handling for editables with content, but where the cursor is on a boundary or new boundary

## Related Issues

- https://github.com/pixiebrix/pixiebrix-extension/pull/7862

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #8157 
- Consolidates draftjs-specific logic
- Always use paste event to insert text into draftjs

## Remaining Work

- [x] Figure out how to interact with draftjs with playwright and complete tests
- [x] Test on other other live draft.js editors: https://github.com/nikgraf/awesome-draft-js?tab=readme-ov-file#standalone-editors-built-on-draftjs

## Demo

- See playwright test

## Checklist

- [x] Add tests and/or storybook stories: playwright tests
- [x] Designate a primary reviewer: @fungairino

### CODE REVIEW

File: end-to-end-tests/tests/runtime/insertAtCursor.spec.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  const modActivationPage = new ActivateModPage(page, extensionId, modId);
+  await modActivationPage.goto();
+
+  await modActivationPage.clickActivateAndWaitForModsPageRedirect();
+
+  await page.goto("/advanced-fields/");
+
+  // The mod contains a trigger to open the sidebar on h1
+  await page.click("h1");
```
Comment (fungairino) [inline end-to-end-tests/tests/runtime/insertAtCursor.spec.ts:35]: We should probably figure out a general utility method for opening the sidebar without relying on a specific mod trigger

File: end-to-end-tests/tests/runtime/insertAtCursor.spec.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  // Need to simulate the mouse entering the sidebar to track focus on MV2
+  // https://github.com/pixiebrix/pixiebrix-extension/blob/1794863937f343fbc8e3a4434eace74191f8dfbd/src/contentScript/sidebarController.tsx#L563-L563
+  const sidebarFrame = page.locator("#pixiebrix-extension");
+
+  if (await sidebarFrame.count()) {
+    await sidebarFrame.dispatchEvent("mouseenter");
+  }
+
+  await sideBarPage.getByRole("button", { name: "Insert at Cursor" }).click();
```
Comment (fungairino) [inline end-to-end-tests/tests/runtime/insertAtCursor.spec.ts:85]: Could we wrap in a if MV2 conditional? Just to make it easier to figure out where the MV2 or MV3 specific logic lies.

**Review by fungairino — APPROVED** [review]

File: end-to-end-tests/tests/runtime/insertAtCursor.spec.ts (hunk 1 above)
Comment (twschiller) [reply end-to-end-tests/tests/runtime/insertAtCursor.spec.ts:35]: 💯  But I don't want it to be a blocker for this PR

File: end-to-end-tests/tests/runtime/insertAtCursor.spec.ts (hunk 2 above)
Comment (twschiller) [reply end-to-end-tests/tests/runtime/insertAtCursor.spec.ts]: Added here: https://github.com/pixiebrix/pixiebrix-extension/pull/8160/files#diff-165a836b26d0898ecd22beeeceb25a9990c96c362be2cca31f7f44bac5764edaR78

File: src/utils/domFieldUtils.ts [hunk 3]
```diff
@@ -108,9 +95,10 @@ export async function setFieldValue(
     // `insertText` acts as a "paste", so if no text is selected it's just appended
     document.execCommand("selectAll");
 
-    if (field.textContent === "" && field.closest(".DraftEditor-root")) {
+    if (field.textContent === "" && isDraftJsField(field)) {
       // Special handling for DraftJS if the field is empty
       // https://github.com/pixiebrix/pixiebrix-extension/issues/7630
+      // XXX: should this just always send a paste command for Draft.js fields?
```
Comment (fregante) [inline src/utils/domFieldUtils.ts:101]: The answer should be yes. If it fails there, it fails here too

File: src/contrib/draftjs/draftJsDom.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * DraftJS doesn't handle `insertText` correctly in most cases, but it can handle this
+ * event. Note that this event doesn't do anything in regular contentEditable fields.
+ * Source: https://github.com/facebookarchive/draft-js/issues/616#issuecomment-426047799
+ */
+export function dispatchPasteForDraftJs(field: HTMLElement, value: string) {
```
Comment (fregante) [inline src/contrib/draftjs/draftJsDom.ts:23]: I'd just rename this to `insertIntoDraftJs`

File: end-to-end-tests/tests/runtime/insertAtCursor.spec.ts (hunk 1 above)
Comment (twschiller) [reply end-to-end-tests/tests/runtime/insertAtCursor.spec.ts:36]: Created an issue here: https://github.com/pixiebrix/pixiebrix-extension/issues/8165

**Pull request merged**